### PR TITLE
Keycloak api key generation provider

### PR DESCRIPTION
**Pull request by gmurtaza00**
# Keycloak custom provider
- This provider creates a unique key upon user registration and set that key as a user-attribute namely api-key.
- The user can login to his account and see the api-key.
- This api-key will be sent by user in header _('x-api-key')_ to access protected routes of the api-key.
- This provider also creates a custom route to verify the api-key
   http://localhost:8080/realms/realm_name/check?apiKey={api-key}

### CODE REVIEW

**Review by slifty — COMMENTED** [review]
This is NOT a full review but wanted to post since I have to go AFK -- main question is whether we can add a linter for Java (does such a thing exist?)

File: api-key-module/pom.xml [hunk 1]
```diff
@@ -0,0 +1,82 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>api-key-module</artifactId>
+    <groupId>com.pdc</groupId>
```
Comment (bickelj) [inline api-key-module/pom.xml:8]: This probably should be either `org.pdc` or `org.philanthropydatacommons` and likewise throughout.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,16 @@
-# keycloak-extensions
+# Keycloak-extensions
 PDC related extensions that were made for the keycloak auth service
+
+## Keycloak installation
+For installation and basic settings of keycloak, follow this document 
+https://docs.google.com/document/d/1aawlITZu_6xAcVetyqvE7NrETay7eAOhVXwfnJcwC7A/edit?usp=sharing
```
Comment (bickelj) [inline README.md:6]: This document helps me see how to get started locally with a keycloak instance, but I am thinking about how to get this same setup programmatically using a bitnami keycloak image without exposing the keycloak interfaces to the world.

Comment (bickelj) [reply README.md:6]: I guess we would expose the keycloak service, right? That way folks can register, regenerate api keys, etc.?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,16 @@
...
+
+## Compiling the custom provider through maven
+
+```mvn -f api-key-module package```
```
Comment (bickelj) [inline README.md:10]: I was able to run this successfully locally within this branch.

File: api-key-module/src/main/java/com/pdc/providers/ApiKeyResourceProviderFactory.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.keycloak.Config;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.services.resource.RealmResourceProvider;
+import org.keycloak.services.resource.RealmResourceProviderFactory;
+
+public class ApiKeyResourceProviderFactory implements RealmResourceProviderFactory {
+
+
```
[the remark this answers is not in the document]
Comment (bickelj) [reply api-key-module/src/main/java/com/pdc/providers/ApiKeyResourceProviderFactory.java:11]: Checkstyle comes with a googly configuration:
```
$ dpkg -L checkstyle
/.
/usr
/usr/bin
/usr/bin/checkstyle
/usr/share
/usr/share/checkstyle
/usr/share/checkstyle/google_checks.xml
```
E.g.
```
keycloak-extensions/api-key-module$ checkstyle -c /usr/share/checkstyle/google_checks.xml src/main/java/com/pdc/providers/*.java
Starting audit...
[WARN] keycloak-extensions/api-key-module/src/main/java/com/pdc/providers/ApiKeyResourceProviderFactory.java:12: 'method def modifier' has incorrect indentation level 4, expected level should be 2. [Indentation]
[WARN] keycloak-extensions/api-key-module/src/main/java/com/pdc/providers/ApiKeyResourceProviderFactory.java:13: 'method def' child has incorrect indentation level 8, expected level should be 4. [Indentation]
```

File: api-key-module/pom.xml (hunk 1 above)
Comment (slifty) [reply api-key-module/pom.xml:8]: FWIW I vote for the more descriptive one (also matches the URL)

File: README.md (hunk 2 above)
Comment (gmurtaza00) [reply README.md:6]: Yeah we will have to expose keycloak service for the purpose of login and register.

**Review by bickelj — CHANGES_REQUESTED** [review]
This is good for the module/jar name but then there are also the package statements in each class at the top, the directory structure leading to those too.

**Review by bickelj — APPROVED** [review]
I do not yet know enough about keycloak or JWT to know the implications of including this jar, but I see that you made the requested changes and this should be merged.

**Comment by jasonaowen** [comment]
Is this PR still relevant? It seems like it may have been written before the plan for OAuth tokens was adopted; is that accurate?

**Comment by bickelj** [comment]
@jasonaowen I would say it is superceded by the use of OAuth 2 / OIDC JWTs, yes.

Looking back at my objections to this API key extension approach, namely the "internal SPI" and "extending keycloak" objections in https://github.com/PhilanthropyDataCommons/service/issues/96#issuecomment-1307619151 the same "internal SPI" warning occurs with PR #3 and it too is an extension to keycloak. I think I am still uneasy with the API key extension, though. The approach in this PR was to add a secret API key to a keycloak user's attributes and then let passing that value be the authentication (or bypass authentication). We implemented a simple version of this without keycloak in the app instead.

**Comment by bickelj** [comment]
@slifty @gmurtaza00 @reefdog If any of y'all think this should be resurrected, please re-open.

**Comment by jasonaowen** [comment]
Thank you for the history, @bickelj! It is very helpful.

**Pull request closed without merging**